src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -809,6 +859,42 @@ private void MoveDownCustomComparer((TElement Element, TPriority Priority) node,
             nodes[nodeIndex] = node;
         }
 
+        /// <summary>
+        /// Scans the heap for the first index containing an element equal to the specified parameter.
+        /// </summary>
+        private int FindIndex(TElement element, IEqualityComparer<TElement>? equalityComparer)
+        {
+            equalityComparer ??= EqualityComparer<TElement>.Default;
+            (TElement Element, TPriority Priority)[] nodes = _nodes;
+            int size = _size;
+
+            // Currently the JIT doesn't optimize direct EqualityComparer<T>.Default.Equals
+            // calls for reference types, so we want to cache the comparer instance instead.
+            // TODO https://github.com/dotnet/runtime/issues/10050: Update if this changes in the future.
+            if (typeof(TElement).IsValueType && equalityComparer == EqualityComparer<TElement>.Default)
+            {
+                for (int i = 0; i < size; i++)
+                {
+                    if (EqualityComparer<TElement>.Default.Equals(element, nodes[i].Element))
```

Comment:
>EqualityComparer<TElement>.Default [](http://example.com/codeflow?start=24&length=34)

does using equalityComparer here will prevent the JIT to optimize it? even for value types?

Reply:
The JIT will devirtualize `EqualityComparer<TElement>.Default.Equals` calls but only when `TElement` is a value type. For reference types it's better to reference the equality comparer from a local.